In [ ]:
import pandas as pd
import numpy as np

# Load Titanic dataset
df = pd.read_csv("titanic.csv")

# Always start by checking the data
print("First 5 rows:")
display(df.head())

print("\nDataset information:")
df.info()

# TASK 01 — THE 15 TITANIC QUESTIONS

# 1. How many passengers are in the dataset?
passenger_count = len(df)

print("\n1. Number of passengers:")
print(passenger_count)


# 2. What fraction of passengers survived overall?
survival_rate = df["survived"].mean()

print("\n2. Overall survival rate:")
print(survival_rate)


# 3. What was the survival rate for women versus men?
survival_by_sex = df.groupby("sex")["survived"].mean()

print("\n3. Survival rate for women versus men:")
print(survival_by_sex)


# 4. What was the survival rate for each passenger class?
survival_by_class = df.groupby("pclass")["survived"].mean()

print("\n4. Survival rate for each passenger class:")
print(survival_by_class)


# 5. What was the average age of passengers?
average_age = df["age"].mean()

print("\n5. Average age of passengers:")
print(average_age)


# 6. Who was the oldest passenger, and who was the youngest?
oldest_passenger = df.loc[df["age"].idxmax()]
youngest_passenger = df.loc[df["age"].idxmin()]

print("\n6. Oldest passenger:")
print(oldest_passenger[["who", "sex", "age"]])

print("\nYoungest passenger:")
print(youngest_passenger[["who", "sex", "age"]])


# 7. What was the average fare paid in each class?
average_fare_by_class = df.groupby("pclass")["fare"].mean()

print("\n7. Average fare paid in each class:")
print(average_fare_by_class)


# 8. How many passengers were children (under 18)?
children = df[df["age"] < 18]
number_of_children = len(children)

print("\n8. Number of children under 18:")
print(number_of_children)


# 9. What was the survival rate among children?
children_survival_rate = children["survived"].mean()

print("\n9. Survival rate among children:")
print(children_survival_rate)


# 10. How many passengers embarked from each port?
passengers_by_port = df["embarked"].value_counts()

print("\n10. Number of passengers from each port:")
print(passengers_by_port)


# 11. How many passengers have a missing age?
missing_age = df["age"].isna().sum()

print("\n11. Number of passengers with missing age:")
print(missing_age)


# 12. What was the survival rate for women in 1st class?
women_first_class = df[
    (df["sex"] == "female") &
    (df["pclass"] == 1)
]

women_first_class_survival = women_first_class["survived"].mean()

print("\n12. Survival rate for women in 1st class:")
print(women_first_class_survival)


# 13. What was the survival rate for men in 3rd class?
men_third_class = df[
    (df["sex"] == "male") &
    (df["pclass"] == 3)
]

men_third_class_survival = men_third_class["survived"].mean()

print("\n13. Survival rate for men in 3rd class:")
print(men_third_class_survival)


# 14. What was the average family size on board?
# Family size = siblings/spouses + parents/children
df["family_size"] = df["sibsp"] + df["parch"]

average_family_size = df["family_size"].mean()

print("\n14. Average family size:")
print(average_family_size)


# 15. Did passengers travelling alone survive at a different rate
# than those with family?

# Define travelling alone using sibsp + parch
df["alone_from_family"] = df["family_size"] == 0

survival_alone_vs_family = df.groupby("alone_from_family")["survived"].mean()

print("\n15. Survival rate: travelling alone vs with family:")
print(survival_alone_vs_family)

First 5 rows:


,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
0,0,3,male,22.0,1,0,7.2500,S,Third,man,True,NaN,Southampton,no,False
1,1,1,female,38.0,1,0,71.2833,C,First,woman,False,C,Cherbourg,yes,False
2,1,3,female,26.0,0,0,7.9250,S,Third,woman,False,NaN,Southampton,yes,True
3,1,1,female,35.0,1,0,53.1000,S,First,woman,False,C,Southampton,yes,False
4,0,3,male,35.0,0,0,8.0500,S,Third,man,True,NaN,Southampton,no,True



Dataset information:
<class 'pandas.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   survived     891 non-null    int64  
 1   pclass       891 non-null    int64  
 2   sex          891 non-null    str    
 3   age          714 non-null    float64
 4   sibsp        891 non-null    int64  
 5   parch        891 non-null    int64  
 6   fare         891 non-null    float64
 7   embarked     889 non-null    str    
 8   class        891 non-null    str    
 9   who          891 non-null    str    
 10  adult_male   891 non-null    bool   
 11  deck         203 non-null    str    
 12  embark_town  889 non-null    str    
 13  alive        891 non-null    str    
 14  alone        891 non-null    bool   
dtypes: bool(2), float64(2), int64(4), str(7)
memory usage: 92.4 KB

1. Number of passengers:
891

2. Overall survival rate:
0.3838383838383838

3. Survival rate for wome

# TASK 01 — Titanic: 15 Questions

## Answers

### 1. How many passengers are in the dataset?
The dataset contains **[result from Q1] passengers**.

### 2. What fraction of passengers survived overall?
The overall survival rate was **[result from Q2]**, meaning this fraction of passengers survived.

### 3. What was the survival rate for women versus men?
The survival rate was **[female result] for women** and **[male result] for men**.

### 4. What was the survival rate for each passenger class?
The survival rates were **[1st class result] for 1st class, [2nd class result] for 2nd class, and [3rd class result] for 3rd class**.

### 5. What was the average age of passengers?
The average passenger age was **[result from Q5] years**.

### 6. Who was the oldest passenger, and who was the youngest?
The oldest passenger was **[name] at [age] years**, while the youngest was **[name] at [age] years**.

### 7. What was the average fare paid in each class?
The average fare was **[1st class result] for 1st class, [2nd class result] for 2nd class, and [3rd class result] for 3rd class**.

### 8. How many passengers were children (under 18)?
There were **[result from Q8] passengers under 18 years old**.

### 9. What was the survival rate among children?
The survival rate among children was **[result from Q9]**.

### 10. How many passengers embarked from each port?
The dataset had **[C result] passengers from C, [Q result] from Q, and [S result] from S**.

### 11. How many passengers have a missing age?
There were **[result from Q11] passengers with a missing age**.

### 12. What was the survival rate for women in 1st class?
The survival rate for women in 1st class was **[result from Q12]**.

### 13. What was the survival rate for men in 3rd class?
The survival rate for men in 3rd class was **[result from Q13]**.

### 14. What was the average family size on board?
The average family size was **[result from Q14] people**, calculated using `sibsp + parch`.

### 15. Did passengers travelling alone survive at a different rate than those with family?
Passengers travelling alone had a survival rate of **[alone result]**, compared with **[family result]** for passengers travelling with family.

---

## Look It Up

### Groupby or filtering?
For questions involving survival rates for groups, `groupby()` is usually convenient because it calculates the result for every group at once. Filtering is useful when we need one specific group or multiple conditions.

### Definition of travelling alone
A passenger is defined as travelling alone when:

`family_size = sibsp + parch = 0`

Therefore, `alone = True` means the passenger had no recorded siblings, spouses, parents, or children travelling with them.

### What does the survival data reveal?
The survival results show that survival was not the same across all passenger groups. Differences can be observed by **sex, passenger class, and whether passengers travelled with family**.

In [ ]:
import pandas as pd
import numpy as np

# Load dataset
df = pd.read_csv("titanic.csv")

# TASK 02 — THREE STRATEGIES FOR MISSING AGE

# 1. Confirm missing age values and fraction
missing_age_count = df["age"].isna().sum()
missing_age_fraction = df["age"].isna().mean()

print("1. Missing age values:", missing_age_count)
print("Fraction of data with missing age:", missing_age_fraction)

# STRATEGY 1 — DROP ROWS WITH MISSING AGE

df_drop = df.dropna(subset=["age"])

drop_mean = df_drop["age"].mean()
drop_std = df_drop["age"].std()
drop_rows = len(df_drop)

print("\n--- Strategy 1: Drop missing ages ---")
print("Rows remaining:", drop_rows)
print("Mean age:", drop_mean)
print("Standard deviation:", drop_std)

# STRATEGY 2 — FILL MISSING AGE WITH OVERALL MEAN

df_mean = df.copy()

overall_mean = df_mean["age"].mean()

df_mean["age"] = df_mean["age"].fillna(overall_mean)

mean_fill_mean = df_mean["age"].mean()
mean_fill_std = df_mean["age"].std()

print("\n--- Strategy 2: Mean fill ---")
print("Mean age:", mean_fill_mean)
print("Standard deviation:", mean_fill_std)

# STRATEGY 3 — FILL USING MEDIAN OF SAME CLASS AND SEX

df_group = df.copy()

df_group["age"] = df_group.groupby(
    ["pclass", "sex"]
)["age"].transform(
    lambda x: x.fillna(x.median())
)

group_fill_mean = df_group["age"].mean()
group_fill_std = df_group["age"].std()

print("\n--- Strategy 3: Group fill ---")
print("Mean age:", group_fill_mean)
print("Standard deviation:", group_fill_std)

# COMPARISON TABLE

comparison = pd.DataFrame({
    "Strategy": [
        "Drop",
        "Mean Fill",
        "Group Fill"
    ],
    "Rows": [
        len(df_drop),
        len(df_mean),
        len(df_group)
    ],
    "Mean Age": [
        drop_mean,
        mean_fill_mean,
        group_fill_mean
    ],
    "Standard Deviation": [
        drop_std,
        mean_fill_std,
        group_fill_std
    ]
})

print("\n--- Comparison of Three Strategies ---")
display(comparison.round(2))

# SURVIVAL RATE FOR WOMEN VS MEN

print("\n--- Survival Rate: Original Dataset ---")
original_survival = df.groupby("sex")["survived"].mean()
print(original_survival)


print("\n--- Survival Rate: Strategy 1 — Drop ---")
drop_survival = df_drop.groupby("sex")["survived"].mean()
print(drop_survival)


print("\n--- Survival Rate: Strategy 2 — Mean Fill ---")
mean_survival = df_mean.groupby("sex")["survived"].mean()
print(mean_survival)


print("\n--- Survival Rate: Strategy 3 — Group Fill ---")
group_survival = df_group.groupby("sex")["survived"].mean()
print(group_survival)

# SURVIVAL RATE COMPARISON TABLE

survival_comparison = pd.DataFrame({
    "Original": original_survival,
    "Drop": drop_survival,
    "Mean Fill": mean_survival,
    "Group Fill": group_survival
})

print("\n--- Survival Rate Comparison ---")
display(survival_comparison.round(4))

1. Missing age values: 177
Fraction of data with missing age: 0.19865319865319866

--- Strategy 1: Drop missing ages ---
Rows remaining: 714
Mean age: 29.69911764705882
Standard deviation: 14.526497332334042

--- Strategy 2: Mean fill ---
Mean age: 29.69911764705882
Standard deviation: 13.002015226002882

--- Strategy 3: Group fill ---
Mean age: 29.11242424242424
Standard deviation: 13.304424074688162

--- Comparison of Three Strategies ---


,Strategy,Rows,Mean Age,Standard Deviation
0,Drop,714,29.70,14.53
1,Mean Fill,891,29.70,13.00
2,Group Fill,891,29.11,13.30



--- Survival Rate: Original Dataset ---
sex
female    0.742038
male      0.188908
Name: survived, dtype: float64

--- Survival Rate: Strategy 1 — Drop ---
sex
female    0.754789
male      0.205298
Name: survived, dtype: float64

--- Survival Rate: Strategy 2 — Mean Fill ---
sex
female    0.742038
male      0.188908
Name: survived, dtype: float64

--- Survival Rate: Strategy 3 — Group Fill ---
sex
female    0.742038
male      0.188908
Name: survived, dtype: float64

--- Survival Rate Comparison ---


,Original,Drop,Mean Fill,Group Fill
sex,,,,
female,0.7420,0.7548,0.7420,0.7420
male,0.1889,0.2053,0.1889,0.1889


# TASK 02 — Three Strategies for Missing Age

## 1. Missing Age

There are **177 missing age values**, which is about **19.87% of the dataset**.

## 2. Strategy 1 — Drop

Rows with missing `age` are removed using `dropna()`, leaving **714 rows**.

The mean age after dropping missing values is approximately **29.70 years**.

## 3. Strategy 2 — Mean Fill

Missing ages are replaced with the overall mean age using `fillna()`.

The mean stays approximately **29.70 years**, while the standard deviation becomes smaller because many missing values are replaced by exactly the same value.

## 4. Strategy 3 — Group Fill

Each missing age is replaced with the **median age of passengers having the same passenger class and sex**.

This is smarter than using one overall mean because age distributions differ between groups such as male/female and 1st/2nd/3rd class.

## 5. Comparison

The comparison table printed above shows the resulting mean age and standard deviation for all three strategies.

**Drop:** removes incomplete observations, so it keeps only passengers with known ages.

**Mean Fill:** keeps all passengers but puts every missing age at the same overall mean, which reduces the observed spread.

**Group Fill:** keeps all passengers while using information from similar passengers, producing a more representative estimate for each missing age.

## 6. Survival Rate — Women vs Men

The survival-rate comparison table shows the results after each missing-age strategy.

The survival rates should be very similar across the three cleaned datasets because **survival rate by sex does not require age to be known**.

This tells us that the choice of age-imputation strategy has little effect on this particular survival comparison.

## Reflection

### Why does filling with a single value shrink standard deviation?

Replacing many different unknown ages with one identical mean value adds no variation around the mean, so the measured spread becomes smaller.

### When would dropping rows be wrong?

Dropping rows can be a problem when many values are missing because it reduces the dataset and may introduce bias if the missing values are not random.

### When is dropping rows fine?

Dropping can be reasonable when only a small number of observations are missing and removing them is unlikely to change the analysis.

### Can imputation change conclusions?

Yes, imputation can change statistical results because estimated values are being added to the dataset, so a poor imputation method can introduce bias and affect conclusions.

### What did this Titanic analysis show?

For the women-versus-men survival comparison, the results change very little across the three age strategies because age is not needed to calculate survival rate by sex.

# TASK 03 — Reflection: Which Strategy, and Why?

For this Titanic dataset, I would choose **Strategy 3 (group fill)** because it keeps all passengers while estimating missing ages using the median of passengers with the same class and sex. Dropping missing ages is simple and avoids making assumptions, but it removes about 20% of the age observations, which could reduce the data available for later analysis and potentially introduce bias. Mean filling keeps all the data, but replacing many different missing ages with one value reduces the natural spread of the age data. Group filling is more representative because age patterns can differ between passenger classes and sexes, although it is still an estimate and is not guaranteed to be correct. There is no single “correct” strategy because the appropriate method depends on the goal; for example, if only complete age records were required for a specific analysis and missingness was small, dropping rows could be preferable. Since about 20% of ages are missing, the imputation method should be clearly documented because hiding that decision could make the analysis misleading and reduce trust in the results.

In [ ]:
import pandas as pd
import numpy as np

# Load Titanic dataset
df = pd.read_csv("titanic.csv")

# TASK 04 — FILTER, SORT, RANK

# 1. Female, 1st or 2nd class, and under 30

filtered_passengers = df[
    (df["sex"] == "female") &
    (df["pclass"].isin([1, 2])) &
    (df["age"] < 30)
]

filtered_count = len(filtered_passengers)
filtered_survival_rate = filtered_passengers["survived"].mean()

print("1. Female, 1st/2nd class, under 30")
print("Number of passengers:", filtered_count)
print("Survival rate:", filtered_survival_rate)

print("\nMatching passengers:")
display(
    filtered_passengers[
        ["sex", "pclass", "age", "fare", "survived"]
    ]
)

# 2. 10 oldest passengers who survived

oldest_survivors = (
    df[df["survived"] == 1]
    .sort_values("age", ascending=False)
    .head(10)
)

print("\n2. 10 oldest passengers who survived:")

display(
    oldest_survivors[
        ["sex", "pclass", "age", "fare"]
    ]
)

# 3. Top 10% vs Bottom 10% fare-payers

top_10_threshold = df["fare"].quantile(0.90)
bottom_10_threshold = df["fare"].quantile(0.10)

top_10_fare = df[df["fare"] >= top_10_threshold]
bottom_10_fare = df[df["fare"] <= bottom_10_threshold]

top_10_survival = top_10_fare["survived"].mean()
bottom_10_survival = bottom_10_fare["survived"].mean()

print("\n3. Fare ranking and survival")

print("\nTop 10% fare threshold:", top_10_threshold)
print("Top 10% number of passengers:", len(top_10_fare))
print("Top 10% survival rate:", top_10_survival)

print("\nBottom 10% fare threshold:", bottom_10_threshold)
print("Bottom 10% number of passengers:", len(bottom_10_fare))
print("Bottom 10% survival rate:", bottom_10_survival)

# 4. Sort by class ascending and fare descending

sorted_df = df.sort_values(
    ["pclass", "fare"],
    ascending=[True, False]
)

print("\n4. Dataset sorted by class ascending and fare descending:")

display(
    sorted_df[
        ["pclass", "fare", "sex", "age", "survived"]
    ].head(20)
)

# 5. Describe fare and age

print("\n5. Descriptive statistics for fare and age:")

display(
    df[["fare", "age"]].describe()
)

# 6. Value counts for passenger class

print("\n6. Passenger class counts:")

print(
    df["pclass"].value_counts().sort_index()
)

# 7. Value counts for embarkation port

print("\n7. Embarkation port counts:")

print(
    df["embarked"].value_counts()
)

# 8. Rank passengers by fare

df["fare_rank"] = df["fare"].rank(
    method="min",
    ascending=False
)

print("\n8. Top 10 passengers ranked by fare:")

display(
    df[
        ["fare", "fare_rank", "pclass", "sex", "survived"]
    ]
    .sort_values("fare_rank")
    .head(10)
)

1. Female, 1st/2nd class, under 30
Number of passengers: 71
Survival rate: 0.9295774647887324

Matching passengers:


,sex,pclass,age,fare,survived
9,female,2,14.0,30.0708,1
41,female,2,27.0,21.0000,0
43,female,2,3.0,41.5792,1
53,female,2,29.0,26.0000,1
56,female,2,21.0,10.5000,1
...,...,...,...,...,...
853,female,1,16.0,39.4000,1
866,female,2,27.0,13.8583,1
874,female,2,28.0,24.0000,1
880,female,2,25.0,26.0000,1



2. 10 oldest passengers who survived:


,sex,pclass,age,fare
630,male,1,80.0,30.0000
275,female,1,63.0,77.9583
483,female,3,63.0,9.5875
570,male,2,62.0,10.5000
829,female,1,62.0,80.0000
587,male,1,60.0,79.2000
366,female,1,60.0,75.2500
195,female,1,58.0,146.5208
268,female,1,58.0,153.4625
11,female,1,58.0,26.5500



3. Fare ranking and survival

Top 10% fare threshold: 77.9583
Top 10% number of passengers: 90
Top 10% survival rate: 0.7666666666666667

Bottom 10% fare threshold: 7.55
Bottom 10% number of passengers: 92
Bottom 10% survival rate: 0.14130434782608695

4. Dataset sorted by class ascending and fare descending:


,pclass,fare,sex,age,survived
258,1,512.3292,female,35.0,1
679,1,512.3292,male,36.0,1
737,1,512.3292,male,35.0,1
27,1,263.0000,male,19.0,0
88,1,263.0000,female,23.0,1
341,1,263.0000,female,24.0,1
438,1,263.0000,male,64.0,0
311,1,262.3750,female,18.0,1
742,1,262.3750,female,21.0,1
118,1,247.5208,male,24.0,0



5. Descriptive statistics for fare and age:


,fare,age
count,891.000000,714.000000
mean,32.204208,29.699118
std,49.693429,14.526497
min,0.000000,0.420000
25%,7.910400,20.125000
50%,14.454200,28.000000
75%,31.000000,38.000000
max,512.329200,80.000000



6. Passenger class counts:
pclass
1    216
2    184
3    491
Name: count, dtype: int64

7. Embarkation port counts:
embarked
S    644
C    168
Q     77
Name: count, dtype: int64

8. Top 10 passengers ranked by fare:


,fare,fare_rank,pclass,sex,survived
679,512.3292,1.0,1,male,1
737,512.3292,1.0,1,male,1
258,512.3292,1.0,1,female,1
341,263.0000,4.0,1,female,1
88,263.0000,4.0,1,female,1
438,263.0000,4.0,1,male,0
27,263.0000,4.0,1,male,0
742,262.3750,8.0,1,female,1
311,262.3750,8.0,1,female,1
118,247.5208,10.0,1,male,0


: 

# TASK 04 — Filter, Sort, Rank

## 1. Multi-condition filter

There are **[run code to get result] female passengers** who were in 1st or 2nd class and under 30, and their survival rate was **[result]**.

`isin([1, 2])` is cleaner than writing two separate OR conditions because it directly means that the class value belongs to the selected group of classes.

## 2. Ten oldest survivors

The table above shows the **10 oldest passengers who survived**, including their sex, passenger class, age, and fare.

## 3. Top 10% vs bottom 10% fare-payers

The top 10% fare-payers had a survival rate of **[result]**, while the bottom 10% had a survival rate of **[result]**.

A higher fare is associated with different survival rates in this dataset, but fare itself may also represent other factors such as passenger class, access to resources, and where passengers were located on the ship, so this comparison does not prove that paying more directly caused survival.

## 4. Sorting

The dataset was sorted first by `pclass` in ascending order and then by `fare` in descending order within each class.

Sorting changes the **order of rows**, while ranking assigns each row a **position or rank** based on a value such as fare.

## 5. Describe — Fare

`describe()` shows the count, mean, standard deviation, minimum, quartiles, and maximum for fare.

**Interesting observation:** the maximum fare is much higher than the median fare, showing that a small number of passengers paid exceptionally high fares.

## 6. Describe — Age

`describe()` summarizes the distribution of passenger ages.

**Interesting observation:** the age column has fewer valid observations than the total number of passengers because some ages are missing.

## 7. Value counts — Passenger class

`value_counts()` shows how many passengers were in each passenger class.

**Interesting observation:** the 3rd-class group contains the largest number of passengers.

## 8. Value counts — Embarkation port

`value_counts()` shows how many passengers embarked from each port.

**Interesting observation:** Southampton (`S`) is the most common embarkation port.

## Key reflection

A **filter** selects rows that satisfy conditions, **sorting** changes their order, and **ranking** assigns positions based on a value; these are different operations even though they can be used together for analysis.